test_fms/parser/generic_blocks.F90

```diff
@@ -0,0 +1,111 @@
+  integer, allocatable :: key_ids(:)         !< The ids of the keys
+  character(len=50)    :: variable_name      !< The variable name
+  character(len=50)    :: model_type_name    !< The model type
+  character(len=50)    :: block_name         !< The name of the block
+  character(len=50)    :: key_name           !< The name of the key
+  character(len=50)    :: key_value          !< The value of the key
+  character(len=50)    :: varnames(2)        !< The expected names of the variables
+  character(len=50)    :: blocknames1(1)     !< The expected names of the blocks for the first variable
+  character(len=50)    :: blocknames2(2)     !< The expected names of the blocks for the second variable
+  character(len=50)    :: keys(5)            !< The expected names of the keys
+  character(len=50)    :: values(5)          !< The expected names values of they keys
+  integer              :: key_count          !< To keep track of the expected answers
+
+  logical :: correct_answer   !< True if the answer is correct
+  integer :: i, j, k, l, m, n !< For do loops
+
+  call fms_init()
+  varnames(1) = "sphum"
+  varnames(2) = "soa"
+
+  blocknames1(1) = "profile_type"
+  blocknames2(1) = "chem_param"
+  blocknames2(2) = "profile_type"
+
+  key_count = 0
+  keys(1) = "value";         values(1) = "fixed"
+  keys(2) = "surface_value"; values(2) = "3.0e-06"
+  keys(3) = "value";         values(3) = "aerosol"
+  keys(4) = "value";         values(4) = "fixed"
+  keys(5) = "surface_value"; values(5) = "1.0e-32"
+
+  yaml_id = open_and_parse_file("sample.yaml")
+  allocate(field_table_ids(get_num_blocks(yaml_id, "field_table")))
+  call get_block_ids(yaml_id, "field_table", field_table_ids)
+  do i = 1, size(field_table_ids)
+    allocate(modlist_ids(get_num_blocks(yaml_id, "modlist", parent_block_id=field_table_ids(i))))
+    call get_block_ids(yaml_id, "modlist", modlist_ids, field_table_ids(i))
+
+    do j = 1, size(modlist_ids)
+      call get_value_from_key(yaml_id, modlist_ids(j), "model_type", model_type_name)
+      print *, "Modlist::", trim(model_type_name)
+      if (trim(model_type_name) .ne. "atmos_mod") &
+        call mpp_error(FATAL, "Modlist is not the expected result")
+
+      allocate(varlist_ids(get_num_blocks(yaml_id, "varlist", parent_block_id=modlist_ids(j))))
+      call get_block_ids(yaml_id, "varlist", varlist_ids, modlist_ids(j))
+
+      do k = 1, size(varlist_ids)
+        call get_value_from_key(yaml_id, varlist_ids(k), "variable", variable_name)
+        print *, "Variable::", trim(variable_name)
+        if (trim(variable_name) .ne. varnames(k)) &
+          call mpp_error(FATAL, "Variable is not the expected result")
+
+        allocate(block_ids(get_num_unique_blocks(yaml_id, parent_block_id=varlist_ids(k))))
+        call get_unique_block_ids(yaml_id, block_ids, parent_block_id=varlist_ids(k))
+        do l = 1, size(block_ids)
+          call get_block_name(yaml_id, block_ids(l), block_name)
```

Comment:
@J-Lentz @GFDL-Eric 
This is relevant code that will be used to read the new field_table yaml format.
`get_num_unique_blocks` is going to be used to get the number of unique blocks for each variable. For the example below, it will get 1 for `sphum` and 2 for `soa`. `get_unique_block_ids` will be used to get an id for each of the blocks and that can be used to get the block name (i.e `profile_type` for `sphum`, and  `chem_param` and `profile_type` for `soa`) and from there, it can be read as normal ...
```
field_table:
- field_type: tracer
  modlist:
  - model_type: atmos_mod
    varlist:
    - variable: sphum
      longname: specific humidity
      units: kg/kg
      profile_type:
      - value: fixed
        surface_value: 3.0e-06
    - variable: soa
      longname: SOA tracer
      units: mmr
      convection: all
      chem_param:
      - value: aerosol
      profile_type:
      - value: fixed
        surface_value: 1.0e-32
```